### PR TITLE
virttest/qemu_vm.py: Add migrate_capabilities parameter for migrate()

### PR DESCRIPTION
**Pull request by spcui**
1. Add set/get migrate_capabilitiy function
2. Add migrate_capabilities parameter for migrate()

Signed-off-by: Shuping Cui <scui@redhat.com>

### CODE REVIEW

File: virttest/qemu_monitor.py [hunk 1]
```diff
@@ -2038,3 +2038,32 @@ def qom_get(self, path, qproperty):
...
+        return self.cmd(cmd, args)
+
+    def get_migrate_capability(self, capability=None):
+        """
+        Get the state of migrate-capability.
+
+        :param capability: capability which need to get.
+        :return: the state of migrate-capability, if capability=None,
+                 will return the list of query-migrate-capabilities.
+        """
```
Comment (xutian) [inline virttest/qemu_monitor.py]: Hi,

Please optimize your logical, function name no match what you do in your function.
I guess it should be:
for info in self.query("migrate-capabilities"):
      if info["capability"] == capability:
           return info["state"]
 return False

**Comment by suqinhuang** [comment]
Hi @spcui could you provide the id?

Regards

File: virttest/qemu_vm.py [hunk 2]
```diff
@@ -3669,6 +3670,15 @@ def migrate(self, timeout=virt_vm.BaseVM.MIGRATE_TIMEOUT, protocol="tcp",
...
                 self.monitor.cmd("stop")
 
+            if migrate_capabilities:
+                for key, value in migrate_capabilities.items():
+                    state = value == "on"
+                    self.monitor.set_migrate_capability(state, key)
+                    s = self.monitor.get_migrate_capability(key)
+                    if s != state:
+                        raise exceptions.TestError("Cannot get the correct "
+                                                   "migrate capability value.")
```
Comment (suqinhuang) [inline virttest/qemu_vm.py]: 1. "Failed to set migrate capability" or "Failed to turn on migrate capability" make more clear.
2. you will set all capabilities? xbzrle, auto-converge ... at the same time?

**Comment by spcui** [comment]
ID: 1297289

File: virttest/qemu_vm.py [hunk 3]
```diff
@@ -3669,6 +3670,15 @@ def migrate(self, timeout=virt_vm.BaseVM.MIGRATE_TIMEOUT, protocol="tcp",
...
                 self.monitor.cmd("stop")
 
+            if migrate_capabilities:
+                for key, value in migrate_capabilities.items():
+                    state = value == "on"
+                    self.monitor.set_migrate_capability(state, key)
+                    s = self.monitor.get_migrate_capability(key)
+                    if s != state:
+                        raise exceptions.TestError("Failed to get "
+                                                   "value of %s." % key)
```
Comment (suqinhuang) [inline virttest/qemu_vm.py]: Suggest to update "get" to "set"

File: virttest/qemu_vm.py [hunk 4]
```diff
@@ -3669,6 +3670,15 @@ def migrate(self, timeout=virt_vm.BaseVM.MIGRATE_TIMEOUT, protocol="tcp",
             if offline is True:
                 self.monitor.cmd("stop")
 
+            if migrate_capabilities:
+                for key, value in migrate_capabilities.items():
```
Comment (suqinhuang) [inline virttest/qemu_vm.py]: suggest to add operation logging info

File: virttest/qemu_vm.py (hunk 3 above)
Comment (spcui) [inline virttest/qemu_vm.py]: no, this is checking get the value.

Comment (xutian) [inline virttest/qemu_vm.py]: Has same question with @suqinguang.   exception looks very strange. better to:
"Migration capability '%s' should be '%s' , but it's '%s'" % ( value, s and "on" or "off")

**Comment by spcui** [comment]
@xutian @suqinhuang updated with your comments, could review again?

**Comment by suqinhuang** [comment]
Acked-by: Suqin Huang

File: virttest/qemu_monitor.py [hunk 5]
```diff
@@ -1147,6 +1147,33 @@ def balloon(self, size):
...
+
+    def get_migrate_capability(self, capability):
+        """
+        Get the state of migrate-capability.
+
+        :param capability: capability which need to get.
+        :return: the state of migrate-capability.
+        """
+        capability_info = self.query("migrate_capabilities")
+        pattern = r"%s:\s(.{2,3})\s" % capability
```
Comment (xutian) [inline virttest/qemu_monitor.py]: why not  r'on|off' ?

**Comment by spcui** [comment]
@xutian updated with your comment, pls review it again.

**Comment by spcui** [comment]
@xutian updated with your comment, pls review it again.

File: virttest/qemu_monitor.py [hunk 6]
```diff
@@ -1147,6 +1147,33 @@ def balloon(self, size):
...
+    def get_migrate_capability(self, capability):
+        """
+        Get the state of migrate-capability.
+
+        :param capability: capability which need to get.
+        :return: the state of migrate-capability.
+        """
+        capability_info = self.query("migrate_capabilities")
+        pattern = r"(?:%s:\s+)on|off" % capability
+        value = re.search(pattern, capability_info, re.M).group(0)
```
Comment (xutian) [inline virttest/qemu_monitor.py]: Please test it, it looks wrong.

Comment (xutian) [inline virttest/qemu_monitor.py]: a = "capabilities: xbzrle: off rdma-pin-all: off auto-converge: off zero-blocks: off events: off"
m = re.search("xbzrle:\s+(on|off)", a)
m.group(1)  == 'on'

Comment (spcui) [inline virttest/qemu_monitor.py]: no wrong here, i'm using r"(?:%s:\s+)on|off" % capability do it, only get on or off.

Comment (xutian) [inline virttest/qemu_monitor.py]: >>> a = "capabilities: xbzrle: on rdma-pin-all: off auto-converge: off zero-blocks: off events: off"
>>> r = re.search("(?:xbzrle:\s+)on|off", a)
>>> r.group(0)
'xbzrle: on'
>>> 
it's my test result.

**Pull request merged**